### DISCUSSION TITLE
Enrich automatic differentiation API - ideas for new fns

### DISCUSSION THREAD

**Discussion by sritchie** [sicmutils/sicmutils#234] [Ideas]
I've been studying forward-mode and reverse-mode AD quite a bit as I work on #221 and #226 , and I think that we have a nice opportunity to build out a rich autodiff API.

This library is a nice home for it because of the work we've started in `sicmutils.expression.compile`; if we can build more expressive derivatives, we can pass symbols through the resulting functions and them compile them into super fast specialized functions.

I WOULD LOVE comments on what these should be named!

### `directional-derivative`, Jacobian-Vector Product

`(D f)` for some `f` that takes multiple args, or a structure, currently returns a Jacobian, ie, a structure of the same dimension `n` as the input with opposite orientation. It does this by calling `derivative` `n` times with a perturbation on each argument.

If, instead, we tagged ALL arguments with the same `tag`, then the result would be a directional derivative, or a Jacobian-vector product.

### Return both `f(x)` and `Df(x)` in forward-mode

We have to calculate both... it'd be nice to have some other operator that returns both things to save the effort if you need both.

### reverse mode: return `f(x)` and a function that WILL return the gradient

Slightly lower level than the eventual API we'll have for reverse mode, which will do this then immediately call the fn.

### Vector-Jacobian Product in reverse-mode

Similar to JVP in forward mode, by supplying a vector of "sensitivities" to start the reverse process, you can calculate a vector-jacobian product in one shot (see the [jax docs](https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html#Vector-Jacobian-products-(VJPs,-aka-reverse-mode-autodiff)) for discussion).

### Mixed-Mode AD, Hessian-Vector product

We can compose a reverse pass with a forward pass to get a two-shot Hessian-vector product (see the [jax docs](https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html#Hessian-vector-products-using-both-forward--and-reverse-mode) for info here). That would be a great example of how to use mixed-mode AD, and a good place to test whether all of this works in nested mode too, by make up some "Alexey's Amazing Bug" style tests for mixed.

### Jet support for forward mode

Right now to get a Jacobian we make `n` calls to `derivative` for a function that takes a structure of dimension `n`. I think we can do this in a single shot (which will still take work linear in the input dimension, but will save a bunch of work over what we're doing now.

The data structure to do this is a [`Jet`](https://github.com/typelevel/spire/blob/master/core/src/main/scala/spire/math/Jet.scala). I think we can use our existing tag system to:

- perturb all the input variables with a different tag, and call the function a single time
- modify `extract-tangent` to pull MANY tags at once (given the original input structure). The rule for a `Jet` here is that any tag in the same group cancels if you multiply it by any other tag; so kill all terms with more than one input tag (where `(> (count (intersection tagset input-tags)) 1)` and then extract each tag like normal
- `rename-tag` needs a new `multi-` arity that can rename a full mapping of tags at once. The rule here should be that:
  -  if the tag set contains ANY of the values of the renaming set, the whole term goes to zero
  - slightly more subtle, if there are any duplicate values, if any tag-set contains BOTH tags, the term has to go to zero too.

Why not just use a `Jet`? A `Jet` can't handle the case of nested derivatives, where non-identical tags can multiply and not annihalate.

### Proper, principled complex number support

We're close here, we just need some principled claims about how complex numbers should be handled. `JAX` does a nice job here and [lays out their approach in their docs](https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html#Complex-numbers-and-differentiation).
